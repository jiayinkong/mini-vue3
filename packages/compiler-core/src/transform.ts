import { NodeTypes } from './ast';
import { TO_DISPLAY_STRING } from './runtimeHelpers';

export function transform(root, options = {}) {
  const context = createTransformText(root, options);

  // 1.遍历 - 深度优先遍历
  traverseNode(root, context);

  createRootCodegen(root);

  root.helpers = [...context.helpers.keys()];
}

function createRootCodegen(root) {
  root.codegenNode = root.children[0];
  const child = root.children[0];
  if(child.type === NodeTypes.ELEMENT) {
    root.codegenNode = child.codegenNode;
  } else {
    root.codegenNode = root.children[0];
  }
}

function createTransformText(root, options) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Map(),
    helper(key) {
      context.helpers.set(key, 1);
    }
  };

  return context;
}

function traverseNode(node, context) {
  const nodeTransforms = context.nodeTransforms;
  let existFns: any = [];

  for(let i = 0; i < nodeTransforms.length; i++) {
    const transform = nodeTransforms[i];
    const onExist = transform(node, context);
    if(onExist) existFns.push(onExist);
  }

  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      context.helper(TO_DISPLAY_STRING);
      break;

    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
      traverseChildren(node, context);
      break;
  
    default:
      break;
  }

  let i = existFns.length;
  while(i--) {
    existFns[i]();
  }
}

function traverseChildren(node, context) {
  const children = node.children;

  for(let i = 0; i < children.length; i++) {
    const node = children[i];
    traverseNode(node, context);
  }
}