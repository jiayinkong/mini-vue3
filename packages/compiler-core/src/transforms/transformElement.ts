import { NodeTypes, createVNodeCall } from '../ast';

export function transforElement(node, context) {
  if(node.type === NodeTypes.ELEMENT) {
    return () => {

      // 中间处理层
      const vnodeTag = `'${node.tag}'`;
      const vnodeChildren = node.children[0];
      let vnodeProps;

      node.codegenNode = createVNodeCall(context, vnodeTag, vnodeProps, vnodeChildren);
    }
  }
}