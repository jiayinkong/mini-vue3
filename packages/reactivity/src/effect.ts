import { extend } from '@mini-vue3/shared';

interface EffectOptions {
  scheduler?: Function;
  onStop?: Function | undefined;
}

let targetMap = new Map();
let activeEffect;
let shouldTrack = false;

export class ReactiveEffect {
  private _fn: Function;
  private active = true;
  public scheduler?: Function;
  public onStop?: Function;
  public deps = [];
  constructor(_fn: Function, scheduler?) {
    this._fn = _fn;
    this.scheduler = scheduler;
  }
  run() {
    if(!this.active) {
      return this._fn();
    }

    // 应该收集
    shouldTrack = true;
    activeEffect = this;
    const r = this._fn();

    // 重置
    shouldTrack = false;

    return r;
  }
  stop() {
    if(this.active) {
      cleanupEffect(this);
      if(this.onStop) {
        this.onStop();
      }
      this.active = false;
    }
  }
}

function cleanupEffect(effect) {
  effect.deps.forEach((dep: any) => {
    dep.delete(effect);
  });
  effect.deps.length = 0;
}

export function track(target, key) {
  if(!isTracking()) return;

  // target -> key -> deps
  let depsMap = targetMap.get(target);

  if(!depsMap) {
    depsMap = new Map();
    targetMap.set(target, depsMap);
  }

  let dep = depsMap.get(key);
  if(!dep) {
    dep = new Set();
    depsMap.set(key, dep);
  }

  trackEffects(dep);
}

export function trackEffects(dep) {
  if(dep.has(activeEffect)) return;

  dep.add(activeEffect);
  activeEffect.deps.push(dep);
}

export function isTracking() {
  return shouldTrack && activeEffect !== undefined;
}

export function trigger(target, key) {
  const depsMap = targetMap.get(target);

  if(!depsMap) {
    return;
  }

  const dep = depsMap.get(key);

  triggerEffects(dep);
}

export function triggerEffects(dep) {
  for(const effect of dep) {
    if(effect.scheduler) {
      effect.scheduler();
    } else {
      effect.run();
    }
  }
}

export function effect(fn, options?: EffectOptions) {
  const _effect = new ReactiveEffect(fn, options?.scheduler);
  extend(_effect, options);
  _effect.run();

  const runner: any  = _effect.run.bind(_effect);
  runner.effect = _effect;
  return runner;
}

export function stop(runner: any) {
  runner.effect.stop();
}