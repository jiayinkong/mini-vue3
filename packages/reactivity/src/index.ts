export { ref, unRef, isRef, proxyRefs } from './ref';
export { reactive, readonly, isReactive, isReadonly, shallowReadonly, isProxy } from './reactive';
export { effect, stop } from './effect';