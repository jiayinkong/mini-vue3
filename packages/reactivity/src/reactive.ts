import { isObject } from '@mini-vue3/shared';
import {
  mutableHanlders,
  readonlyHandlers,
  shallowReadonlyHandlers,
} from "./basehandlers";

export const enum ReactiveFlags {
  IS_REACTIVE = "__v_isReative",
  IS_READONLY = "__v_isReadonly",
}

export function reactive(raw) {
  return createReactiveObject(raw, mutableHanlders);
}

export function isReactive(value) {
  return !!value[ReactiveFlags.IS_REACTIVE];
}

export function readonly(raw) {
  return createReactiveObject(raw, readonlyHandlers);
}

export function isReadonly(value) {
  return !!value[ReactiveFlags.IS_READONLY];
}

export function shallowReadonly(raw) {
  return createReactiveObject(raw, shallowReadonlyHandlers);
}

export function isProxy(value) {
  return isReactive(value) || isReadonly(value);
}

function createReactiveObject(target, baseHandler) {
  if(!isObject(target)) {
    console.warn(`target ${target} should be a object`);
  }
  return new Proxy(target, baseHandler);
}
