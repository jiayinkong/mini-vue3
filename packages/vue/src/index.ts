export * from '@mini-vue3/runtime-dom';

import { baseCompile } from '@mini-vue3/compiler-core';
import * as runtimeDom from '@mini-vue3/runtime-dom';
import { registerRuntimeCompiler } from '@mini-vue3/runtime-dom';

function compileToFunction(template) {
  const { code } = baseCompile(template);
  const render = new Function('Vue', code)(runtimeDom);
  return render;
}

registerRuntimeCompiler(compileToFunction);


          // compiler-core
// Vue ->     
          // runtime-dom -> runtime-core -> reactivity