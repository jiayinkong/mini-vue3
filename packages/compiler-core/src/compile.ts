import { generate } from './codegen';
import { baseParse } from './parse';
import { transform } from './transform';
import { transforElement } from './transforms/transformElement';
import { transformExpression } from './transforms/transformExpression';
import { transformText } from './transforms/transformText';

export function baseCompile(template) {
  const ast = baseParse(template);

  transform(ast, {
    nodeTransforms: [transformExpression, transforElement, transformText]
  });

  return generate(ast);
}