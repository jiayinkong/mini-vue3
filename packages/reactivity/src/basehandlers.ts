import { extend, isObject } from '@mini-vue3/shared';
import { track, trigger } from './effect';
import { ReactiveFlags, reactive, readonly } from './reactive';

const get = createGetter();
const set = createSetter();
const readonlyGet = createGetter(true);
const shallowReadonlyGet = createGetter(true, true);

function createGetter(isReadonly = false, shallow = false) {
  return function get(target, key) {
    if(key === ReactiveFlags.IS_REACTIVE) {
      return !isReadonly;
    }
    if(key === ReactiveFlags.IS_READONLY) {
      return isReadonly;
    }
    const res = Reflect.get(target, key);

    if(isReadonly && shallow) {
      return res;
    }

    // 看看 res 是不是 object
    if(isObject(res)) {
      return isReadonly ? readonly(res) : reactive(res);
    }

    if(!isReadonly) {
      // 依赖收集
      track(target, key);
    }

    return res;
  }
}

function createSetter() {
  return function set(target, key, value) {
    const res = Reflect.set(target, key, value);
    
    // 触发依赖
    trigger(target, key);
    return res;
  }
}

export const mutableHanlders = {
  get,
  set,
}

export const readonlyHandlers = {
  get: readonlyGet,
  set(target, key, value) {
    console.warn('不允许修改');
    return true;
  }
}

export const shallowReadonlyHandlers = extend({}, readonlyHandlers, {
  get: shallowReadonlyGet
})
