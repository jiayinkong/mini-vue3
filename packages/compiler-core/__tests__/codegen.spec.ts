import { generate } from '../src/codegen';
import { baseParse } from '../src/parse'
import { transform } from '../src/transform';
import { transforElement } from '../src/transforms/transformElement';
import { transformExpression } from '../src/transforms/transformExpression';
import { transformText } from '../src/transforms/transformText';

describe('generate', () => {
  it('string', () => {
    const ast = baseParse('hiiii');

    transform(ast);

    const { code } = generate(ast);
  
    expect(code).toMatchSnapshot();
  });

  it('inperpolation', () => {
    const ast = baseParse('{{message}}');

    transform(ast, {
      nodeTransforms: [transformExpression]
    });

    const { code } = generate(ast);
  
    expect(code).toMatchSnapshot();
  });

  it('element', () => {
    const ast: any = baseParse('<div>Hi, {{message}}</div>');

    transform(ast, {
      nodeTransforms: [transformExpression, transforElement, transformText]
    });

    const { code } = generate(ast);
  
    expect(code).toMatchSnapshot();
  });
})