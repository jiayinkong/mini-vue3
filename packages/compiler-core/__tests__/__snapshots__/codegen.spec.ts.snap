// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generate > element 1`] = `
"const { toDisplayString:  _toDisplayString, createElementVNode:  _createElementVNode } = Vue
return function render(_ctx, _cache){return _createElementVNode('div', null, 'Hi, ' + _toDisplayString(_ctx.message))}"
`;

exports[`generate > inperpolation 1`] = `
"const { toDisplayString:  _toDisplayString } = Vue
return function render(_ctx, _cache){return _toDisplayString(_ctx.message)}"
`;

exports[`generate > string 1`] = `
"
return function render(_ctx, _cache){return 'hiiii'}"
`;

exports[`generate element 1`] = `
"const { toDisplayString:  _toDisplayString, createElementVNode:  _createElementVNode } = Vue
return function render(_ctx, _cache){return _createElementVNode('div', null, 'Hi, ' + _toDisplayString(_ctx.message))}"
`;

exports[`generate inperpolation 1`] = `
"const { toDisplayString:  _toDisplayString } = Vue
return function render(_ctx, _cache){return _toDisplayString(_ctx.message)}"
`;

exports[`generate string 1`] = `
"
return function render(_ctx, _cache){return 'hiiii'}"
`;
